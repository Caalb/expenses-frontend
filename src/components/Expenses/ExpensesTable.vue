<template>
  <QCard bordered flat>
    <QTable
      :rows="expenses"
      :columns="columns"
      row-key="id"
      :loading="loading"
      :hide-pagination="true"
    >
      <template v-slot:body="props">
        <QTr :props="props">
          <QTd key="description" :props="props">
            {{ props.row.description }}
          </QTd>

          <QTd key="date" :props="props">
            {{ formatDate(props.row.date) }}
          </QTd>

          <QTd key="value" :props="props">
            {{ formatCurrency(props.row.value) }}
          </QTd>

          <QTd key="actions" :props="props">
            <div class="row items-center q-gutter-sm justify-end">
              <QBtn
                flat
                round
                color="primary"
                icon="edit"
                size="sm"
                @click="$emit('edit', props.row)"
              >
                <QTooltip>Editar</QTooltip>
              </QBtn>

              <QBtn
                flat
                round
                color="negative"
                icon="delete"
                size="sm"
                @click="$emit('delete', props.row)"
              >
                <QTooltip>Excluir</QTooltip>
              </QBtn>
            </div>
          </QTd>
        </QTr>
      </template>

      <template v-slot:no-data>
        <div class="full-width row flex-center q-pa-md text-grey-6">
          Nenhuma despesa encontrada
        </div>
      </template>
    </QTable>
  </QCard>
</template>

<script setup lang="ts">
import { QCard, QTable, QTr, QTd, QBtn, QTooltip } from 'quasar'
import { format } from 'date-fns'
import { ptBR } from 'date-fns/locale'
import type { Expense } from '@/types/expense'

defineProps<{
  expenses: Expense[]
  loading: boolean
}>()

defineEmits<{
  (e: 'edit', expense: Expense): void
  (e: 'delete', expense: Expense): void
}>()

const columns = [
  {
    name: 'description',
    required: true,
    label: 'Descrição',
    align: 'left' as const,
    field: 'description'
  },
  {
    name: 'date',
    required: true,
    label: 'Data',
    align: 'left' as const,
    field: 'date'
  },
  {
    name: 'value',
    required: true,
    label: 'Valor',
    align: 'left' as const,
    field: 'value'
  },
  {
    name: 'actions',
    required: true,
    label: 'Ações',
    align: 'right' as const,
    field: 'actions'
  }
]

const formatDate = (date: string) => {
  return format(new Date(date), 'dd/MM/yyyy', { locale: ptBR })
}

const formatCurrency = (value: number) => {
  return new Intl.NumberFormat('pt-BR', {
    style: 'currency',
    currency: 'BRL'
  }).format(value)
}
</script>
