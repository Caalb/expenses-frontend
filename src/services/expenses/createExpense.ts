import type { Expense } from '@/types/expense'
import api from '@/services/api'

export const createExpense = async (expense: Omit<Expense, 'id'>) => {
  const response = await api.post('/expenses', expense)

  return response
}
