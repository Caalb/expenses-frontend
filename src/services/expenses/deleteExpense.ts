import api from '@/services/api'

export const deleteExpense = async (id: number) => {
  const response = await api.delete(`/expenses/${id}`)

  return response
}
