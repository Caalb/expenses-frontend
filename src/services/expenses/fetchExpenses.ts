import api from '@/services/api'
import type { Expense } from '@/types/expense'

export const fetchExpenses = async (): Promise<Expense[]> => {
  const response = await api.get<Expense[]>('/expenses')

  return response.data
}
