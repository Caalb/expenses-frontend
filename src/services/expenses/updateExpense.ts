import type { Expense } from '@/types/expense'
import api from '@/services/api'

export const updateExpense = async (id: number, expense: Omit<Expense, 'id'>) => {
  const response = await api.put(`/expenses/${id}`, expense)

  return response
}
