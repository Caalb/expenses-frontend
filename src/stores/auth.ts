import { defineStore } from 'pinia'
import { userSignIn } from '../services/auth/userSignIn'
import type { AxiosResponse } from 'axios'
import { userSignUp } from '@/services/auth/userSignUp'

export const useAuthStore = defineStore('auth', () => {
  const handleSignIn = async (
    email: string,
    password: string,
  ): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignIn(email, password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  const handleSignUp = async (data: Record<string, string>): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignUp(data.name, data.email, data.password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  return {
    handleSignIn,
    handleSignUp,
  }
})
