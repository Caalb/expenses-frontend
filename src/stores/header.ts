import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useHeaderStore = defineStore('header', () => {
  const activeSubmenu = ref<string | null>(null)

  const setActiveSubmenu = (submenu: string) => {
    activeSubmenu.value = activeSubmenu.value === submenu ? null : submenu
  }

  const clearSubmenu = () => {
    activeSubmenu.value = null
  }

  return {
    activeSubmenu,
    setActiveSubmenu,
    clearSubmenu,
  }
})
