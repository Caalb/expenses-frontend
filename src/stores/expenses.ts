import { createExpense as createExpenseApi } from '@/services/expenses/createExpense'
import { fetchExpenses as fetchExpensesApi } from '@/services/expenses/fetchExpenses'
import type { Expense } from '@/types/expense'
import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useExpensesStore = defineStore('expenses', () => {
  const expenses = ref<Expense[]>([])

  const fetchExpenses = async () => {
    try {
      const response = await fetchExpensesApi()

      expenses.value = response

      return response
    } catch (error) {
      return error
    }
  }

  const createExpense = async (expense: Omit<Expense, 'id'>) => {
    try {
      const response = await createExpenseApi(expense)
      expenses.value = [response.data, ...expenses.value]

      return response
    } catch (error) {
      return error
    }
  }

  return {
    expenses,
    fetchExpenses,
    createExpense,
  }
})
