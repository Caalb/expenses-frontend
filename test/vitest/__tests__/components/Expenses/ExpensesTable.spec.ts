import ExpensesTable from '@/components/Expenses/ExpensesTable.vue'
import { useExpensesStore } from '@/stores/expenses'
import { createTestingPinia } from '@pinia/testing'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { describe, expect, it, beforeEach, vi } from 'vitest'
import { getByTestId } from 'test/vitest/__utils__'

installQuasarPlugin()

const createWrapper = () =>
  mount(ExpensesTable, {
    global: {
      plugins: [
        createTestingPinia({
          createSpy: vi.fn,
        }),
      ],
    },
  }) as VueWrapper

describe('ExpensesTable', () => {
  let wrapper: VueWrapper
  let expensesStore: ReturnType<typeof useExpensesStore>

  beforeEach(() => {
    wrapper = createWrapper()
    expensesStore = useExpensesStore()
    expensesStore.expenses = [
      { id: 1, amount: '100', date: '2023-01-01', description: 'Expense 1' },
      { id: 2, amount: '200', date: '2023-01-02', description: 'Expense 2' },
      { id: 3, amount: '300', date: '2023-01-03', description: 'Expense 3' },
    ]
  })

  it('should display expenses when store has data', async () => {
    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit edit when edit button is clicked', async () => {
    await wrapper.vm.$nextTick()

    const editButton = getByTestId(wrapper, 'edit-button')
    expect(editButton.exists()).toBe(true)

    await editButton.trigger('click')

    expect(wrapper.emitted('edit')).toBeTruthy()
    expect(wrapper.emitted('edit')?.[0]).toEqual([
      { id: 1, amount: '100', date: '2023-01-01', description: 'Expense 1' },
    ])
  })

  it('should emit delete when edit button is clicked', async () => {
    await wrapper.vm.$nextTick()

    const deleteButton = getByTestId(wrapper, 'delete-button')
    expect(deleteButton.exists()).toBe(true)

    await deleteButton.trigger('click')

    expect(wrapper.emitted('delete')).toBeTruthy()
    expect(wrapper.emitted('delete')?.[0]).toEqual([
      { id: 1, amount: '100', date: '2023-01-01', description: 'Expense 1' },
    ])
  })
})

describe('ExpensesTable without expenses', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    wrapper = createWrapper()
  })

  it('should display no expenses message when store is empty', async () => {
    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toContain('Nenhuma despesa encontrada')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
