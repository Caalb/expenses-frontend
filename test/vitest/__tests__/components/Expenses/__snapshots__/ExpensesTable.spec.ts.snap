// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExpensesTable > should display expenses when store has data 1`] = `
"<div class="q-card q-card--bordered q-card--flat no-shadow">
  <div class="q-table__container q-table--horizontal-separator column no-wrap q-table__card q-table--no-wrap">
    <!---->
    <div class="q-table__middle scroll">
      <table class="q-table">
        <thead>
          <tr>
            <th class="text-left">Descrição</th>
            <th class="text-left">Data</th>
            <th class="text-left">Valor</th>
            <th class="text-center" style="width: 100px;">Ações</th>
          </tr>
        </thead>
        <tbody>
          <tr class="q-tr ">
            <td class="q-td text-left">Expense 1</td>
            <td class="q-td text-left">31/12/2022</td>
            <td class="q-td text-left">R$&nbsp;100,00</td>
            <td class="q-td text-center">
              <div class="row items-center q-gutter-sm justify-end"><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-primary q-btn--actionable q-focusable q-hoverable" style="font-size: 10px;" tabindex="0" type="button" data-testid="edit-button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">edit</i><!----></span></button><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-negative q-btn--actionable q-focusable q-hoverable" style="font-size: 10px;" tabindex="0" type="button" data-testid="delete-button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">delete</i><!----></span></button></div>
            </td>
          </tr>
          <tr class="q-tr ">
            <td class="q-td text-left">Expense 2</td>
            <td class="q-td text-left">01/01/2023</td>
            <td class="q-td text-left">R$&nbsp;200,00</td>
            <td class="q-td text-center">
              <div class="row items-center q-gutter-sm justify-end"><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-primary q-btn--actionable q-focusable q-hoverable" style="font-size: 10px;" tabindex="0" type="button" data-testid="edit-button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">edit</i><!----></span></button><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-negative q-btn--actionable q-focusable q-hoverable" style="font-size: 10px;" tabindex="0" type="button" data-testid="delete-button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">delete</i><!----></span></button></div>
            </td>
          </tr>
          <tr class="q-tr ">
            <td class="q-td text-left">Expense 3</td>
            <td class="q-td text-left">02/01/2023</td>
            <td class="q-td text-left">R$&nbsp;300,00</td>
            <td class="q-td text-center">
              <div class="row items-center q-gutter-sm justify-end"><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-primary q-btn--actionable q-focusable q-hoverable" style="font-size: 10px;" tabindex="0" type="button" data-testid="edit-button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">edit</i><!----></span></button><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-negative q-btn--actionable q-focusable q-hoverable" style="font-size: 10px;" tabindex="0" type="button" data-testid="delete-button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">delete</i><!----></span></button></div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
  </div>
</div>"
`;

exports[`ExpensesTable without expenses > should display no expenses message when store is empty 1`] = `
"<div class="q-card q-card--bordered q-card--flat no-shadow">
  <div class="q-table__container q-table--horizontal-separator column no-wrap q-table__card q-table--no-wrap">
    <!---->
    <div class="q-table__middle scroll">
      <table class="q-table">
        <thead>
          <tr>
            <th class="text-left">Descrição</th>
            <th class="text-left">Data</th>
            <th class="text-left">Valor</th>
            <th class="text-center" style="width: 100px;">Ações</th>
          </tr>
        </thead>
        <tbody></tbody>
      </table>
    </div>
    <div class="q-table__bottom row items-center q-table__bottom--nodata">
      <div class="full-width row flex-center q-pa-md text-grey-6"> Nenhuma despesa encontrada </div>
    </div>
  </div>
</div>"
`;
